src/cluster.py

```diff
@@ -0,0 +1,226 @@
+            # Check that the cluster is up and running.
+            if not self._is_cluster_running():
+                raise ClusterNotRunningError()
+
+    def inhibit_default_cluster_creation(self) -> None:
+        """Stop the PostgreSQL packages from creating the default cluster."""
+        os.makedirs(os.path.dirname(CREATE_CLUSTER_CONF_PATH), mode=0o755, exist_ok=True)
+        with open(CREATE_CLUSTER_CONF_PATH, mode="w") as file:
+            file.write("create_main_cluster = false\n")
+            file.write(f"include '{self.conf_path}/conf.d/postgresql-operator.conf'")
+
+    def _create_cluster(self, password: str) -> None:
+        """Create a new PostgreSQL cluster."""
+        # Write the password in a temporary file for pg_createcluster.
+        with open("/tmp/pwfile", mode="w") as file:
+            file.write(password)
+        try:
+            command = [
+                "pg_createcluster",
+                self.version,
+                "main",
+                "--datadir=/var/lib/postgresql/data/pgdata",
+                "--",
+                "--pwfile=/tmp/pwfile",
+            ]
+            logger.info(f"pg_createcluster call: {' '.join(command)}")
+            subprocess.check_call(command)
+        except subprocess.CalledProcessError as e:
+            raise ClusterCreateError(e.stdout)
+        finally:
+            # Remove the temporary file.
+            os.remove("/tmp/pwfile")
+
+    def _is_cluster_running(self) -> bool:
+        """Check whether the cluster is running or not."""
+        try:
+            clusters = (
+                subprocess.check_output(["pg_lsclusters", "--no-header"]).decode().splitlines()
+            )
+
+            def is_online(cluster: str):
+                """Return whether or not a cluster is online."""
+                if cluster.split()[3] == "online":
+                    return True
+                return False
+
+            online_clusters = list(filter(is_online, clusters))
+            return len(online_clusters) == 1
+        except subprocess.CalledProcessError:
+            raise
+
+    def _render_file(self, path: str, content: str, mode: int) -> None:
+        """Write a content rendered from a template to a file."""
+        # Write the content to the file.
+        with open(path, "w+") as file:
+            file.write(content)
+        # Ensure correct permissions are set on the file.
+        os.chmod(path, mode)
+        # Get the uid/gid for the postgres user.
+        u = pwd.getpwnam("postgres")
+        # Set the correct ownership for the file.
+        os.chown(path, uid=u.pw_uid, gid=u.pw_gid)
+
+    def _render_pg_hba_conf_file(self) -> None:
+        """Render the application auth configuration file."""
+        # Open the template pg_hba.conf file.
+        with open("templates/pg_hba.conf.j2", "r") as file:
+            template = Template(file.read())
+        # Define the client authentication rules.
+        connections = [
```

Comment:
Is it intended that these will change at some point? Since they're all constants currently, doesn't seem like much of a need to provide them as variables to the template itself rather than just adding the contents to the file directly.

Reply:
Probably not. Accesses by relation will be made using the last rule. Therefore, the current file content is enough. I'll change to a file with all the content to copy to the file in the unit instead of the template.

Reply:
I removed the template and created a file with all the contents to be copied to the server instance.